lf-queue/lfq.c

```diff
@@ -0,0 +1,223 @@
+            node->next = (void *) -1;
+            free(node); /* we got the lock; actually free */
+            atomic_store(&ctx->is_freeing, false);
+            smb();
+        } else /* we did not get the lock; only add to a freelist */
+            insert_pool(ctx, node);
+    } else
+        insert_pool(ctx, node);
+    free_pool(ctx, false);
+}
+
+static int alloc_tid(struct lfq_ctx *ctx)
+{
+    for (int i = 0; i < ctx->MAX_HP_SIZE; i++) {
+        if (ctx->tid_map[i] == 0) {
+            int old = 0;
+            if (CAS(&ctx->tid_map[i], &old, 1))
+                return i;
+        }
+    }
+
+    return -1;
+}
+
+static void free_tid(struct lfq_ctx *ctx, int tid)
+{
+    ctx->tid_map[tid] = 0;
+}
+
+int lfq_init(struct lfq_ctx *ctx, int max_consume_thread)
+{
+    struct lfq_node *tmp = calloc(1, sizeof(struct lfq_node));
+    if (!tmp)
+        return -errno;
+
+    struct lfq_node *node = calloc(1, sizeof(struct lfq_node));
+    if (!node)
+        return -errno;
+
+    tmp->can_free = node->can_free = true;
+    memset(ctx, 0, sizeof(struct lfq_ctx));
+    ctx->MAX_HP_SIZE = max_consume_thread;
+    ctx->HP = calloc(max_consume_thread, sizeof(struct lfq_node));
+    ctx->tid_map = calloc(max_consume_thread, sizeof(struct lfq_node));
+    ctx->head = ctx->tail = tmp;
+    ctx->fph = ctx->fpt = node;
+
+    return 0;
+}
+
+long lfg_count_freelist(const struct lfq_ctx *ctx)
+{
+    long count = 0;
+    for (struct lfq_node *p = (struct lfq_node *) ctx->fph; p; p = p->free_next)
+        count++;
+    return count;
+}
+
+int lfq_release(struct lfq_ctx *ctx)
+{
+    if (ctx->tail && ctx->head) {               /* if we have data in queue */
+        while ((struct lfq_node *) ctx->head) { /* while still have node */
+            struct lfq_node *tmp = (struct lfq_node *) ctx->head->next;
+            safe_free(ctx, (struct lfq_node *) ctx->head);
+            ctx->head = tmp;
+        }
+        ctx->tail = 0;
+    }
+    if (ctx->fph && ctx->fpt) {
+        free_pool(ctx, true);
+        if (ctx->fph != ctx->fpt)
+            return -1;
+        free(ctx->fpt); /* free the empty node */
+        ctx->fph = ctx->fpt = 0;
+    }
+    if (ctx->fph || ctx->fpt)
+        return -1;
+
+    free(ctx->HP);
+    free(ctx->tid_map);
+    memset(ctx, 0, sizeof(struct lfq_ctx));
+
+    return 0;
+}
+
+int lfq_enqueue(struct lfq_ctx *ctx, void *data)
+{
+    struct lfq_node *insert_node = calloc(1, sizeof(struct lfq_node));
+    if (!insert_node)
+        return -errno;
+
+    insert_node->data = data;
+    struct lfq_node *old_tail = XCHG(&ctx->tail, insert_node);
+    /* We have claimed our spot in the insertion order by modifying tail.
+     * we are the only inserting thread with a pointer to the old tail.
+     *
+     * Now we can make it part of the list by overwriting the NULL pointer in
+     * the old tail. This is safe whether or not other threads have updated
+     * ->next in our insert_node.
+     */
+#ifdef DEBUG
```

Comment:
It is not necessary to have `DEBUG` here. Simply write down `assert`.

Reply:
Quote from https://man7.org/linux/man-pages/man3/assert.3.html
> If the macro NDEBUG is defined at the moment <assert.h> was last included, the macro assert() generates no code, and hence does nothing at all.

Reply:
Done!